openmc/cmfd.py

```diff
@@ -0,0 +1,3072 @@
+                data = matrix.data[matrix.indptr[row]:matrix.indptr[row+1]]
+                for i in range(len(cols)):
+                    fh.write('{:3d}, {:3d}, {:0.8f}\n'.format(
+                        row, cols[i], data[i]))
+
+        # Save matrix in scipy format
+        sparse.save_npz(base_filename, matrix)
+
+    def _calc_fission_source(self):
+        """Calculates CMFD fission source from CMFD flux. If a coremap is
+        defined, there will be a discrepancy between the spatial indices in the
+        variables ``phi`` and ``nfissxs``, so ``phi`` needs to be mapped to the
+        spatial indices of the cross sections. This can be done in a vectorized
+        numpy manner or with for loops
+
+        """
+        # Extract number of groups and number of accelerated regions
+        nx, ny, nz, ng = self._indices
+        n = self._mat_dim
+
+        # Compute cmfd_src in a vecotorized manner by phi to the spatial
+        # indices of the actual problem so that cmfd_flux can be multiplied by
+        # nfissxs
+
+        # Calculate volume
+        vol = np.product(self._hxyz, axis=3)
+
+        # Reshape phi by number of groups
+        phi = self._phi.reshape((n, ng))
+
+        # Extract indices of coremap that are accelerated
+        idx = self._accel_idxs
+
+        # Initialize CMFD flux map that maps phi to actual spatial and
+        # group indices of problem
+        cmfd_flux = np.zeros((nx, ny, nz, ng))
+
+        # Loop over all groups and set CMFD flux based on indices of
+        # coremap and values of phi
+        for g in range(ng):
+            phi_g = phi[:,g]
+            cmfd_flux[idx + (g,)] = phi_g[self._coremap[idx]]
+
+        # Compute fission source
+        cmfd_src = (np.sum(self._nfissxs[:,:,:,:,:] *
+                    cmfd_flux[:,:,:,:,np.newaxis], axis=3) *
+                    vol[:,:,:,np.newaxis])
+
+        # Normalize source such that it sums to 1.0
+        self._cmfd_src = cmfd_src / np.sum(cmfd_src)
+
+        # Compute entropy
+        if openmc.lib.settings.entropy_on:
+            # Compute source times log_2(source)
+            source = self._cmfd_src[self._cmfd_src > 0] \
+                * np.log(self._cmfd_src[self._cmfd_src > 0])/np.log(2)
+
+            # Sum source and store
+            self._entropy.append(-1.0 * np.sum(source))
+
+        # Normalize source so average is 1.0
+        self._cmfd_src = self._cmfd_src/np.sum(self._cmfd_src) * self._norm
+
+        # Calculate differences between normalized sources
+        self._src_cmp.append(np.sqrt(1.0 / self._norm
+                             * np.sum((self._cmfd_src - self._openmc_src)**2)))
+
+    def _cmfd_reweight(self, new_weights):
+        """Performs weighting of particles in source bank
+
+        Parameters
+        ----------
+        new_weights : bool
+            Whether to reweight particles or not
+
+        """
+        # Compute new weight factors
+        if new_weights:
```

Comment:
Question: It looks like `_cmfd_reweight` is inert if `new_weights` is `False`.  If this is true, I might argue that having the `new_weights` parameter may not be worth it.  It also looks like `_cmfd_reweight` is called in only one spot, with `new_weights` being `True`.

If you want to keep the `new_weights` parameter, it may make sense to just `return` if `new_weights` is `False`, and save you a scope (tab indent) for the bulk of the function.